hr_timesheet_purchase_order/models/hr_timesheet_recurrence.py

```diff
@@ -0,0 +1,351 @@
+        ]
+    )
+
+    mon = fields.Boolean(string="Mon")
+    tue = fields.Boolean(string="Tue")
+    wed = fields.Boolean(string="Wed")
+    thu = fields.Boolean(string="Thu")
+    fri = fields.Boolean(string="Fri")
+    sat = fields.Boolean(string="Sat")
+    sun = fields.Boolean(string="Sun")
+
+    repeat_day = fields.Selection([(str(i), str(i)) for i in range(1, 32)])
+    repeat_week = fields.Selection(
+        [
+            ("first", "First"),
+            ("second", "Second"),
+            ("third", "Third"),
+            ("last", "Last"),
+        ]
+    )
+    repeat_weekday = fields.Selection(
+        [
+            ("mon", "Monday"),
+            ("tue", "Tuesday"),
+            ("wed", "Wednesday"),
+            ("thu", "Thursday"),
+            ("fri", "Friday"),
+            ("sat", "Saturday"),
+            ("sun", "Sunday"),
+        ],
+        string="Day Of The Week",
+        readonly=False,
+    )
+    repeat_month = fields.Selection(
+        [
+            ("january", "January"),
+            ("february", "February"),
+            ("march", "March"),
+            ("april", "April"),
+            ("may", "May"),
+            ("june", "June"),
+            ("july", "July"),
+            ("august", "August"),
+            ("september", "September"),
+            ("october", "October"),
+            ("november", "November"),
+            ("december", "December"),
+        ]
+    )
+
+    @api.constrains("repeat_unit", "mon", "tue", "wed", "thu", "fri", "sat", "sun")
+    def _check_recurrence_days(self):
+        for timesheet in self.filtered(lambda p: p.repeat_unit == "week"):
+            if not any(
+                [
+                    timesheet.mon,
+                    timesheet.tue,
+                    timesheet.wed,
+                    timesheet.thu,
+                    timesheet.fri,
+                    timesheet.sat,
+                    timesheet.sun,
+                ]
+            ):
+                raise ValidationError(_("You should select a least one day"))
+
+    @api.constrains("repeat_interval")
+    def _check_repeat_interval(self):
+        if self.filtered(lambda t: t.repeat_interval <= 0):
+            raise ValidationError(_("The interval should be greater than 0"))
+
+    @api.constrains("repeat_number", "repeat_type")
+    def _check_repeat_number(self):
+        if self.filtered(lambda t: t.repeat_type == "after" and t.repeat_number <= 0):
+            raise ValidationError(_("Should repeat at least once"))
+
+    @api.constrains("repeat_type", "repeat_until")
+    def _check_repeat_until_date(self):
+        today = fields.Date.today()
+        if self.filtered(lambda t: t.repeat_type == "until" and t.repeat_until < today):
+            raise ValidationError(_("The end date should be in the future"))
+
+    @api.constrains(
+        "repeat_unit", "repeat_on_month", "repeat_day", "repeat_type", "repeat_until"
+    )
+    def _check_repeat_until_month(self):
+        if self.filtered(
```

Comment:
This filter is too complex and doesn't have tests, please add test case on this behavior.